src/later.cpp

```diff
@@ -204,8 +204,15 @@ bool execCallbacksOne(
     if (callbacks.size() == 0) {
       break;
     }
+
+#ifdef RCPP_USING_UNWIND_PROTECT // See https://github.com/r-lib/later/issues/191
+    // This line may throw errors!
+    callbacks[0]->invoke();
```

Comment:
I attempted to implement this in the latest commit here in my later branch: https://github.com/shikokuchuo/later/commits/uwp/

The strange thing is, it appears to work fine both errors and interrupts when I use this with a mirai / Shiny stack (mirai sends stdFunction callbacks to later). And yet it reliably segfaults during tests / R CMD check.

Reply:
An alternative would be to disable Rcpp unwind protect to revert to the old behaviour (also suggested by Lionel). The safest way is probably by adding the flags to Makevars.in and Makevars.win as I've done in the latest commit here: https://github.com/shikokuchuo/later/commits/disable_uwp/ Note that I've had to skip 2 of the old interrupt tests for clean check results.

Reply:
Making some more progress to help get this over the line:

1. I can confirm that we do not need to `unwindProtect()` the `Rcpp::Function` invocations as these call `unwindProtect()` anyway: https://github.com/RcppCore/Rcpp/blob/6193de530785f1d84117a852cf7e478b1f048b2f/inst/include/Rcpp/Function.h#L82 and https://github.com/RcppCore/Rcpp/blob/6193de530785f1d84117a852cf7e478b1f048b2f/inst/include/Rcpp/api/meat/Rcpp_eval.h#L61-L63.

2. I've modified our `std::function<void(void)>` case to exactly mirror that for `std::function<SEXP(void)>` present in Rcpp's source code at https://github.com/RcppCore/Rcpp/blob/6193de530785f1d84117a852cf7e478b1f048b2f/inst/include/Rcpp/unwindProtect.h#L80. This still segfaults for `R CMD check`, but interestingly does not for `devtools::test()`, although there are 7 test failures (the same failures that are present without my changes on this branch).

My `uwp` branch has been updated: https://github.com/shikokuchuo/later/commits/uwp/.

Reply:
This is all super helpful. @shikokuchuo other than disabling unwind protect, have you found a usage of `unwindProtect()` that fixes my example here? https://github.com/r-lib/later/pull/192#issuecomment-2438598770

Reply:
@lionel- the comment to your original commit to Rcpp is along the lines of what we discussed today that we should unwind protect arbitrary code invocatons: https://github.com/RcppCore/Rcpp/commit/995f17b7db4b5b8cd2dcc8b700b9587c25ef5d68#diff-91c5b46dc84a94604a4e4d0caed9bf85590a2eddbb12d2e8dc80badf324a9dfbR2

However, this is the output from running `R CMD check --use-valgrind` on my latest `uwp` branch: https://gist.github.com/shikokuchuo/a5e4396b4d389276902f732d69a0790d

Specifically looking at `by 0x12DFFEBF: _later_execCallbacks.cold (RcppExports.cpp:130)`
That RcppExports line 130 is actually the `END_RCPP` macro. Specifically, given the `exception_to_r_condition` call, it is this branch of the `END_RCPP` TryCatch: https://github.com/RcppCore/Rcpp/blob/6193de530785f1d84117a852cf7e478b1f048b2f/inst/include/Rcpp/macros/macros.h#L76. So it seems a `std::exception` is being caught rather than a `Rcpp::LongjumpException`. Perhaps this is where we want to get to, after the long jump is unwound, but then getting the last call information somehow leads to the invalid read.

Valgrind actually passes on *this* branch (rather than my `uwp` branch with the modifications from today).

Reply:
> This is all super helpful, thank you both. @shikokuchuo other than disabling unwind protect, have you found a usage of `unwindProtect()` that fixes my example here? [#192 (comment)](https://github.com/r-lib/later/pull/192#issuecomment-2438598770)

Do you mean for 4.1 and earlier on Windows 32-bit? The example works using your branch here on R4.4 Ubuntu and R4.2 on Windows.

Reply:
@shikokuchuo Yes, sorry, I did mean for R 4.1 and earlier on Windows 32-bit.

Reply:
The tests that are failing are the last batch of C++ tests, where we construct R functions that call Rcpp code, which in turn call later on one of these functions that error in various ways :)

It turns out that this may be too much to wrap in `Rcpp::unwindProtect()` - specifically calling into the Rcpp function may be the issue.

To confirm that C code is behaving as we expect (and confirming the anecdotal evidence that mirai/Shiny seems to be working fine) - I skip the C++ tests, and add one of my own that uses nanonext and abuses an internal function that causes a long jump during an `Rf_eval()` from the later callback invocation. Test results here: https://github.com/shikokuchuo/later/actions/runs/11935403862/job/33267183990. Notably it passes for R4.1 on Windows uniformly for both 64bit and 32bit.

So the remaining to do is to investigate why `Rcpp::unwindProtect()` around an Rcpp function isn't safe.

Reply:
In my hacking last night I was just barely able to get the C++ test (`throw std::runtime_error("This is a C++ exception.");`) to pass on 32 bit, by putting `BEGIN_CPP`/`END_CPP` inside `oof()`. (Or maybe I had to introduce a wrapper function to get it to compile, but with that general idea).

Anyway, that worked fine, but I replaced `BEGIN_CPP`/`END_CPP` with `unwindProtect`, expecting it to do the same thing, but the latter did not work.

Reply:
Amazing! I do believe we've done it the 3 of us combined.

I had tried putting these macros around the unwindProtect. But of course it's the other way that works. And the reason it does is the `RCPP_END` macro converts all C++ exceptions to R errors, which are then caught by the unwindProtect (as Lionel succinctly put it earlier in the day). Without that, it seems the exceptions were crossing this boundary and causing the segfault.

I've hence wrapped the `std::function` invocation with these macros, before passing to `Rcpp::unwindProtect`. I've also kept all this logic in one place by using a lambda function, which is explicitly supported, in this commit here: https://github.com/shikokuchuo/later/commit/4e9bf3375834688c6927337e01c765665a255b01

This unfortunately still leaves us with the win32 errors, but given the advantages in terms of (i) long-jump protection for C code and (ii) benefits to [avoiding R_TopLevelExec](https://github.com/r-lib/later/pull/192/files#diff-563cfce67d7007e226622761b341eea58237a2c48b84f9f7b42d4248793b85edR132),  I think it's a good trade-off to make. I'll have a word with Lionel in the morning just to see if we can think of why the exceptions don't seem to be propagated up the stack specifically for that platform.

Reply:
Ohhhh, that makes so much sense when you lay it out like that. Sounds great.

If the Windows x86 problem isn't new, then I don't have a problem shipping this at all. Thanks!

Reply:
Great! I ran this past Lionel and another way to look at it is: "If you're using the unprotected Rcpp callback in an Rcpp stack, all is fine. But by using unwindProtect you introduce an R stack in between. So it should be treated as an R/Rcpp boundary, and the RCPP macros are made for that." He's just great at explanations.

For Win x86 it's never worked AFAICT. I also tried later 1.3.2 and 1.3.0 on an old windows laptop to confirm that it also fails there. Throwing exceptions across module boundaries just doesn't work with Windows SEH (Structured Exception Handling).

On the same topic, Lionel also brought this to my attention: https://stackoverflow.com/questions/13990892/throwing-c-exceptions-outside-static-library If we're catching C++ exceptions thrown by user C++ code, that's also not guaranteed to be safe 100% of the time. But it mostly is if the same compilers etc. are being used. It's also what `later` has always been doing -  using the RCPP macros instead of our own tryCatch block is in no sense worse in this respect.

Whilst it might be worth suggesting in the docs that users can wrap their C++ functions with the RCPP macros for safety (and also ensures it works on Win x86), we've never actually made any requirements on *user* code and I wouldn't want to make that call here. Also looking through this issues, I don't think this has ever come up. In any case, I tested that the current code works fine if users add their own extra layer of `BEGIN_RCPP`/`END_RCPP` in their callback functions.

Reply:
Oh that's interesting. Because C++ code would always include `later_api.h`. Practically though I don't think it makes much difference to just asking users to use BEGIN_RCPP/END_RCPP. If we could define a class that does that with RAII it would save one line, but it'd still need to be instantiated within each function the user writes.

It's a cognitive burden either way. My thoughts were that if it hasn't surfaced for the time `later` has been in existence then it's probably not a major concern?